operator/misc/publisher.go

```diff
@@ -0,0 +1,178 @@
+	TLSEnabled bool
+	Topic string
+}
+
+type MessageEnclosure struct {
+	App string
+	Timestamp string
+	ClusterName string
+	ClusterVersion string
+	Payload interface{}
+
+}
+
+func NewPublisher() *Publisher {
+	pub := &Publisher{}
+	conf := NewKafkaConfig()
+
+	pub.Topic = conf.Topic
+	pub.Writer = &kafka.Writer{
+		Addr:		kafka.TCP(conf.BrokerEndpoint),
+		Async:		false,
+		Transport:	MakeKafkaTransport(conf),
+		Completion: func(messages []kafka.Message, err error) {
+
+		},
+	}
+
+	return pub
+}
+
+func NewKafkaConfig() KafkaConfig {
+	conf := KafkaConfig{
+		Username: GetEnvValue(fmt.Sprintf("%s_PUBLISHER_KAFKA_USERNAME", CONF_PREFIX), ""),
+		Password: GetEnvValue(fmt.Sprintf("%s_PUBLISHER_KAFKA_PASSWORD", CONF_PREFIX), ""),
+		BrokerEndpoint: GetEnvValue(fmt.Sprintf("%s_PUBLISHER_KAFKA_ENDPOINT", CONF_PREFIX), ""),
+		Topic: GetEnvValue(fmt.Sprintf("%s_PUBLISHER_KAFKA_TOPIC", CONF_PREFIX), ""),
+		SASLMechanism: GetEnvValue(fmt.Sprintf("%s_PUBLISHER_KAFKA_SASLMECHANISM", CONF_PREFIX), ""),
+		TLSEnabled: GetEnvBool(fmt.Sprintf("%s_PUBLISHER_KAFKA_TLSENABLED", CONF_PREFIX), false),
+	}
+
+	return conf
+}
+
+func MakeKafkaTransport(conf KafkaConfig) *kafka.Transport {
+	var saslMechanism sasl.Mechanism = nil
+	var tlsConf *tls.Config = nil
+
+	// Support for SCRAM can be added here.
+	switch confSasl := strings.ToLower(conf.SASLMechanism); confSasl {
+	case "plain":
+		saslMechanism = plain.Mechanism{
+			Username: conf.Username,
+			Password: conf.Password,
+		}
+	}
+
+	if conf.TLSEnabled {
+		tlsConf = &tls.Config{}
+	}
+
+	dialer := &kafka.Dialer{
+		Timeout: 10 * time.Second,
+		SASLMechanism: saslMechanism,
+		TLS: tlsConf,
+	}
+
+	netDialer := &net.Dialer{
+		Timeout: dialer.Timeout,
+		Deadline: dialer.Deadline,
+		LocalAddr: dialer.LocalAddr,
+		DualStack: dialer.DualStack,
+		FallbackDelay: dialer.FallbackDelay,
+		KeepAlive: dialer.KeepAlive,
+	}
+
+	transport := &kafka.Transport{
+		Dial: netDialer.DialContext,
+		SASL: dialer.SASLMechanism,
+		TLS: dialer.TLS,
+		ClientID: dialer.ClientID,
+		IdleTimeout: 9 * time.Minute, // See segmentio/kafka-go/writer.go for why this value was chosen
+		MetadataTTL: 15 * time.Second, // See segmentio/kafka-go/writer.go for why this value was chosen
+	}
+
+	return transport
+}
+
+type PublisherService struct {
+	publisher *Publisher
+	messageChannel <-chan interface{}
+}
+
+func NewPublisherService(messageChannel <-chan interface{}) *PublisherService {
+	ps := &PublisherService{
+		publisher: NewPublisher(),
+		messageChannel: messageChannel,
+	}
+
+	return ps
+}
+
+func (p *PublisherService) Run() {
+	conf, err := config.GetConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	clientSet, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		log.Fatal(err)
+	}
+	serverVer, err := clientSet.ServerVersion()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for {
+		msg := <-p.messageChannel
+
+		go func() {
+			enclosure := MessageEnclosure{
+				App:            "inventa-operator",
```

Comment:
Could we perhaps retrieve/determine this name from the operator/deployment rather than hard-code?

Reply:
Could do one of these:
- https://stackoverflow.com/a/37261036
- https://stackoverflow.com/a/37253342

Any particular preference? (I'm fine with either myself)

Reply:
`/etc/hostname` gets the job done quickest I assume, even though there is the potential for it to be `localhost` in some environments. The [pod info](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/) method might be technically better but more complex

I'm happy with either as well, we can always change it later